### PR TITLE
Advanced configuration option validation.

### PR DESCRIPTION
**Pull request by michelle192837**
- Validate that names don't have forbidden prefixes (e.g. can't start with the prefixes of special files we use in TestGrid)
- Validate TestGroup and DashboardTab-specific options

/assign fejta@, chases2@

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -225,6 +207,234 @@ func validateReferencesExist(c configpb.Configuration) error {
 	return mErr.ErrorOrNil()
 }
 
+// validateName validates an entity name is non-empty and contains no prefix that overlaps with a
+// TestGrid file prefix, post-normalization.
+func validateName(s string, entity string) error {
+	name := normalize(s)
+	if name == "" {
+		return &ConfigError{s, entity, "Normalized name can't be empty."}
```
Comment (fejta) [inline config/config.go]: a) How are we using these strongly typed errors?
b) Why do we need to return a reference?

File: config/config.go [hunk 2]
```diff
@@ -225,6 +207,234 @@ func validateReferencesExist(c configpb.Configuration) error {
...
+		if err != nil {
+			mErr = multierror.Append(mErr, &ConfigError{dt.Name, "DashboardTab", err.Error()})
+		}
+	}
+
+	return mErr.ErrorOrNil()
+}
+
+func validateEntityConfigs(c configpb.Configuration) error {
+	mErr := &multierror.Error{}
```
Comment (fejta) [inline config/config.go]: `var mErr error`

https://github.com/hashicorp/go-multierror#usage

File: config/config.go [hunk 3]
```diff
@@ -225,6 +207,234 @@ func validateReferencesExist(c configpb.Configuration) error {
...
+
+	// At the moment, don't need to further validate Dashboards or DashboardGroups.
+	for _, tg := range c.TestGroups {
+		err := validateTestGroup(tg)
+		if err != nil {
+			mErr = multierror.Append(mErr, err)
```
Comment (fejta) [inline config/config.go]: `mErr = multierror.Append(mErr, err)`

File: config/yamlcfg/yaml2proto_test.go [hunk 4]
```diff
@@ -499,7 +521,10 @@ test_groups:
 			input: config.Configuration{
 				TestGroups: []*config.TestGroup{
 					{
-						Name: "test_group",
+						Name:             "test_group",
+						GcsPrefix:        "fake path",
```
Comment (fejta) [inline config/yamlcfg/yaml2proto_test.go]: FYI: I'm not super wild about having so many fields unrelated to "basic dashboard values work" as it makes these cases harder to understand what we're actually trying to test (not GcsPrefix for sure). I don't have a great solution though off-hand

File: config/config.go (hunk 3 above)
Comment (michelle192837) [reply config/config.go]: Sorry, can't tell from this comment, what's the suggestion? (Just append without checking err != nil?, or something else?)

File: config/config.go (hunk 1 above)
Comment (michelle192837) [reply config/config.go]: Most of what I want is to easily display all the errors present in the config to the end user when validating config, and make it clear what parts of the config need to be fixed (i.e. what type + name errored). I'm not sure if these kinds of errors are best or if I should just switch to structured logging at this point?

File: config/yamlcfg/yaml2proto_test.go (hunk 4 above)
Comment (michelle192837) [reply config/yamlcfg/yaml2proto_test.go]: Yeeeeeah, I agree. I was thinking of breaking out a default TestGroup to add to the tests, but I'm not sure if the tradeoff on explicitness is worth it here.

File: config/config.go (hunk 2 above)
Comment (michelle192837) [reply config/config.go]: Has to be a multierror error for ErrorOrNil, but changed.

File: config/yamlcfg/yaml2proto_test.go (hunk 4 above)
Comment (michelle192837) [reply config/yamlcfg/yaml2proto_test.go]: Changed, might be better?

File: config/config.go (hunk 2 above)
Comment (fejta) [reply config/config.go]: I don't see why all this code can't just be the more idiomatic:
```golang
var merr error
if err := foo(); err != nil {
  merr = multierror.Append(merr, err)
}
if err := bar(); err != nil {
  merr = multierror.Append(merr, err)
}
// etc
return merr
```

Comment (michelle192837) [reply config/config.go]: Ah, misunderstood. Done!

File: config/config.go (hunk 3 above)
Comment (michelle192837) [reply config/config.go]: Think this is addressed now.

File: config/config.go (hunk 1 above)
Comment (michelle192837) [reply config/config.go]: Refactored a bit to create ConfigErrors in fewer places.

**Pull request merged**